sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -553,8 +553,12 @@ case class DataSource(
         disallowWritingIntervals(data.schema.map(_.dataType), forbidAnsiIntervals = true)
         SaveIntoDataSourceCommand(data, dataSource, caseInsensitiveOptions, mode)
       case format: FileFormat =>
+        val shouldAllowEmptySchema =
+          sparkSession.sessionState.conf.getConf(SQLConf.ALLOW_EMPTY_SCHEMAS_FOR_WRITES)
         disallowWritingIntervals(data.schema.map(_.dataType), forbidAnsiIntervals = false)
-        DataSource.validateSchema(data.schema)
+        if (!shouldAllowEmptySchema) {
```

Comment:
Should we only honor this config for formats that support it? I'm leaning towards keeping it simple and not checking that here because it is more complicated to maintain as format support changes, but wanted to call it out for discussion.